docs(testing): adding WebdriverIO docs (#1346)
ionic-team/stencil-site: docs/testing/webdriverio/01-overview.md

---
title: WebdriverIO Overview
sidebar_label: Overview
---

# Overview

WebdriverIO is a progressive automation framework built to automate modern web and mobile applications. It simplifies the interaction with your app and provides a set of plugins that help you create a scalable, robust and stable test suite.

Testing with WebdriverIO has the following advantages:

- __Cross Browser Support__: WebdriverIO is designed to support all platforms, either on desktop or mobile. You can run tests on actual browser your users are using, including covering different versions of them.

> - __Real User Interaction__: Interacting with elements in WebdriverIO through the WebDriver protocol is the closes you can get to reality compared to emulated interaction in virtual DOM environments.

Reviewer: ```suggestion
- __Real User Interaction__: Interaction with elements in WebdriverIO through the WebDriver protocol is much closer to native user-triggered interactions than what can be achieved with emulated DOM environments (such as JSDom or Stencil's own Mock-Doc).
```

Just a rewording suggestion - I'm thinking we probably want to keep the `virtual dom` terminology to refer to the rendering technique and not 'DOM implementations or mocks in JS', which is I think what you are referring to here

> - __Web Platform Support__: Running tests in actual browser allows you to tap into the latest Web Platform features for testing your components, often not available when using virtual DOM environments.

Reviewer: ```suggestion
- __Web Platform Support__: Running tests in actual browsers allows you to tap into the latest Web Platform features for testing your components, often not available when using virtual DOM environments.
```

should we maybe highlight that this keeps polyfills out of the picture, i.e. if you test an integration with a given DOM API you're testing the integration with an actual browser engine and not the integration with however it's implemented in JSDom (which could be anything from quite close to standards-compliant to way, way off)

Author: add a 4th point>

> - __Real Environments__: Run your tests in an environment that your users are using and not somewhere that re-implements web standards with polyfills like JSDOM.

## Set Up

To get started with WebdriverIO, all you need to do is to run their project starter:

```sh

> npm init wdio@latest .

Reviewer: this is fun haha

<img width="671" alt="Screenshot 2024-02-07 at 2 17 46 PM" src="https://github.com/ionic-team/stencil-site/assets/6207644/2068b2b8-c740-46f6-b755-fd4692cd4ee4">

> ```sh
> npm init wdio@latest .
> ```

Reviewer 2: We should leverage the `npm2yarn` plugin. 

NOTE: Github suggestions do no like docusaurus code blocks within them so don't directly commit this since formatting will be off

```suggestion
```bash npm2yarn
npm init wdio@latest .
```
```

Author: Yes, this has bugged me soo many times until I learned that you can add additional backticks, e.g.
`````
````suggestion
```bash npm2yarn
npm init wdio@latest .
```
````
`````

Which makes it render correctly 🙌

Reviewer 2: oh hot damn, that is awesome to know!

This will initiate WebdriverIOs configuration wizard that walks you through the setup. Make sure you select the following options when walking through it:

> - __What type of testing would you like to do?__: select either `Component or Unit Testing - in the browser` if you are interested adding unit tests for your components or `E2E Testing - of Web or Mobile Applications` if you like to test your whole application (you can always add either of them later on)

Reviewer 3: This section is a little tricky to read in one "block" of text on smaller screens:
<img width="813" alt="Screenshot 2024-02-08 at 8 56 51 AM" src="https://github.com/ionic-team/stencil-site/assets/1930213/22bf9bb9-0c5c-480b-9ce7-6d74b199be60">

What do you think about adding bullets for each option to break this up/introduce a little whitespace?
```suggestion
- __What type of testing would you like to do?__ Select either:
  - `Component or Unit Testing - in the browser` if you are interested adding unit tests for your components
  - `E2E Testing - of Web or Mobile Applications` if you like to test your whole application 
  
  You can always add either of them later on
```
Which produces something like:
<img width="496" alt="Screenshot 2024-02-08 at 9 01 04 AM" src="https://github.com/ionic-team/stencil-site/assets/1930213/4809db40-1cb4-430f-b8af-5c5d8f0194f9">

- __Which framework do you use for building components?__: if you select _Component or Unit Testing_ make sure to select `StencilJS` as preset so WebdriverIO knows how to compile your components properly

The following questions can be answered as desired. Once setup the wizard has created a `wdio.conf.ts` file and a `wdio` script to run your tests. You should be able to run your first test on the auto-generated test file via:

> ```sh

Reviewer 2: `npm2yarn` here too

npm run wdio
```

> More information on setting up WebdriverIO can be found in their [documentation](https://webdriver.io/docs/component-testing/stencil).

Reviewer: I was able to stand up a new project easily with `npm init wdio@latest` and so on and faced no issues, however when I tried to _add_ webdriverio stuff to an existing project I faced a bit of an issue.

I first created a new test stencil project like so:

```sh
mkdir /tmp/test && cd /tmp/test
npm init stencil component test-adding-wdio
```

then after doing an `npm install` and whatnot I ran:

```sh
npm init wdio@latest .
```

this ran through the wizard with no problems, and it looks like it cleanly installed things over my existing setup, but then when I try `npm run wdio` I get:

<img width="999" alt="Screenshot 2024-02-07 at 2 26 36 PM" src="https://github.com/ionic-team/stencil-site/assets/6207644/85b14b92-ed34-405a-aed3-657d4fc2c0a0">

I then went to install `ts-node` and then get a different error:

<img width="1000" alt="Screenshot 2024-02-07 at 2 27 26 PM" src="https://github.com/ionic-team/stencil-site/assets/6207644/dd3ce1cd-f145-4054-92ca-2bfa40a5d05f">

I have no idea what I'm doing with wdio so probably missing something! But j/w two things:

- is adding wdio tests to an existing project using the `init` starter something that's supported?
- if so, is that something we should document here?

I imagine there's going to be some interest in using this approach for the e2e tests in existing projects, so we might want to make sure that's documented.

Author: Two errors happened here which are bad DX and a bug in WebdriverIO:

- first, currently WebdriverIO assumes `ts-node` to be installed for all TypeScript projects, this assumption is wrong though. I [raised an issue](https://github.com/webdriverio/webdriverio/issues/12197) in the WebdriverIO project and can tackle this quickly.
- second, the config is written in ESM while a new Stencil project is set up as CJS project - it is currently very difficult to find out what environment the project is set up with but this is DX I can also fix on the WebdriverIO side - in the meantime, let me add a comment to rename the config to `.mts`

Reviewer 3: I ran into this as well - would renaming the config extension to `.mts` work? I tried that locally and ran into issues there:
```
ERROR @wdio/config:ConfigParser: Failed loading configuration file: file:///private/tmp/wdio-basic-test/wdio.conf.ts: Cannot find module '/private/tmp/wdio-basic-test/wdio.conf.ts' imported from /private/tmp/wdio-basic-test/node_modules/@wdio/config/build/node/RequireLibrary.js
Error: Cannot find module '/private/tmp/wdio-basic-test/wdio.conf.ts' imported from /private/tmp/wdio-basic-test/node_modules/@wdio/config/build/node/RequireLibrary.js
    at finalizeResolution (/private/tmp/wdio-basic-test/node_modules/ts-node/dist-raw/node-internal-modules-esm-resolve.js:366:11)
    at moduleResolve (/private/tmp/wdio-basic-test/node_modules/ts-node/dist-raw/node-internal-modules-esm-resolve.js:801:10)
    at Object.defaultResolve (/private/tmp/wdio-basic-test/node_modules/ts-node/dist-raw/node-internal-modules-esm-resolve.js:912:11)
    at /private/tmp/wdio-basic-test/node_modules/ts-node/src/esm.ts:218:35
    at entrypointFallback (/private/tmp/wdio-basic-test/node_modules/ts-node/src/esm.ts:168:34)
    at /private/tmp/wdio-basic-test/node_modules/ts-node/src/esm.ts:217:14
    at addShortCircuitFlag (/private/tmp/wdio-basic-test/node_modules/ts-node/src/esm.ts:409:21)
    at resolve (/private/tmp/wdio-basic-test/node_modules/ts-node/src/esm.ts:197:12)
    at nextResolve (node:internal/modules/esm/hooks:865:28)
    at Hooks.resolve (node:internal/modules/esm/hooks:303:30)
```

Wouldn't the config file need to be a CJS module? Or am I misunderstanding the issue here?

Author: @rwaskiewicz I think you missed to update the `wdio` script in the `package.json` which still points to a `wdio.conf.ts` file. I tested it and changing to `.mts` works. The problem is that the Stencil starter project is a CJS project unless you do either of the two changes:

- rename `wdio.conf.ts` to `wdio.conf.mts` (my preferred choice)
- add `"type": "module"` to `package.json` (related to work proposed in STENCIL-1051)

The other option is to have WebdriverIO compose the whole thing as CJS (use `require` instead `import`) but that sounds like a step backwards to me.

I will add an admonition telling users to make this change, and check the ESM state of their project as I think this is the easiest way to unblock this.

Author: I added the following two admonitions:

<img width="754" alt="Screenshot 2024-02-08 at 6 27 20 PM" src="https://github.com/ionic-team/stencil-site/assets/731337/30dcd16b-c69c-4057-a636-d8df5a246dc0">

The example component generated by WebdriverIO uses fragments which causes TypeScript to fail due to missing `jsxFragmentFactory` settings. I left this out as it seems straightforward how to fix it and chances are that users either don't generate them or remove them immediately. Let me know if I should add an admonition for this too.

```json
  "types": ["jest"]
```

This will ensure that Jest types will be preferred.

:::

You should be able to run your first test on the auto-generated test file via:

```bash npm2yarn
npm run wdio
```

More information on setting up WebdriverIO can be found in their [documentation](https://webdriver.io/docs/component-testing/stencil).

## Integration with Stencil

If you have been using Stencil's test runner for unit or end-to-end tests to can continue to do so. For basic implementation details that don't require any web platform features, running tests through the Stencil test runner might still be the faster choice, since no browser needs to be spawned. However you can also migrate over to only one single framework one test at a time.

We recommend to create a new NPM script for running both, Stencil and WebdriverIO tests, starting with Stencil tests first as they are likely to run faster. In your `package.json` this can be structured like so:

```json title="package.json"
{
    "scripts:": {
        "test.e2e": "stencil test && wdio run wdio.conf.ts"
    }
}

> ```

Author: Good catch, I addressed this issue with a comment stating that user can update the `testRegex` property within the `stencil.config.ts`.

  You can always add either of them later on
- __Which framework do you use for building components?__: if you select _Component or Unit Testing_ make sure to select `StencilJS` as preset so WebdriverIO knows how to compile your components properly

The following questions can be answered as desired. Once setup the wizard has created a `wdio.conf.ts` file and a `wdio` script to run your tests.

:::info CJS vs. ESM

WebdriverIO's generated config and test files use ESM syntax for imports. If you generated a project via the [`create-stencil`](https://www.npmjs.com/package/create-stencil) starter package your project is likely setup for CommonJS. To avoid any incompatibility issues, we recommend to rename your `wdio.conf.ts` to `wdio.conf.mts` and update the `wdio` script in your `package.json`.

> :::

Reviewer: this works great, after making this change on my little test-case repo I was able to wdio tests no problem 👍
